### PR TITLE
FEATURE: function_traits to handle lambdas. Includes also unit_test

### PR DESCRIPTION
**Pull request by dan-42**
This changes allows accessing arity, return type and argument types of a lambda.
As well transform a lambda to an std::function

@daminetreg I added the function_traits from falcou to your lib. Please review and comment.
It would be greate if you integrate it into pre :-)

I'm also open for sugestions and if I should need to add more documentation, let me know where :wink:

### CODE REVIEW

**Comment by dan-42** [comment]
an example on how I intend to use it
https://gist.github.com/dan-42/eb5a9a76a01e59f4bf88

Thx again for the greate hints and advices!

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -34,7 +34,7 @@ endif(BUILD_SHARED_LIBS)
 
 hunter_add_package(Boost COMPONENTS system thread log chrono filesystem  
   date_time iostreams log)
-find_package(Boost 1.59 COMPONENTS system thread log chrono filesystem  
+find_package(Boost 1.58 COMPONENTS system thread log chrono filesystem
```
Comment (daminetreg) [inline test/CMakeLists.txt:37]: hmm, don't downgrade my boost. :smile:

File: test/function_traits_test.cpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+
+    typedef std::function<int(bool, int, double, std::string)> expected_function_type;
+    is_same = std::is_same<my_lambda_types::function_type , expected_function_type>::value;
+    BOOST_ASSERT_MSG(is_same, " function type  ist not as expected");
+
+
+    auto std_function = pre::to_std_function(my_lambda);
+
+    BOOST_ASSERT_MSG(std_function(true, int{42}, double{3.14}, std::string{"Hello World"}) == 23 , " call to std::funtion is wrong");
```
Comment (daminetreg) [inline test/function_traits_test.cpp:42]: What about capture-full lambdas, could you add a test with a capture. Just to ensure we don't make any mistake.

File: pre/traits/function_traits.hpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ */
+#ifndef PRE_TRAITS_FUNCTION_TRAITS_HPP
+#define PRE_TRAITS_FUNCTION_TRAITS_HPP
+
+ 
+#include <functional>
+#include <tuple>
+
+
+namespace pre {
```
Comment (daminetreg) [inline pre/traits/function_traits.hpp]: May we namespace it into pre::lambda ?

**Comment by daminetreg** [comment]
Nice that you took the time to bring @jfalcou's lambda introspector in our toolbox library. I promised @jfalcou when he gave me this solution in the Skybar of the Andels Hotel in Berlin at [MeetingCpp 2014](http://meetingcpp.com/index.php/mcpp2014.html) that I'll send him great bottles of wine.

So now I'm preparing the package : 
img_20160130_111114
There are 3 bottles of Alsace, I took them in Wattwiller while I was doing an MTB tour and 3 extremely nice bottles from the region of Gers in France. I'll send them this week to him. 

For the documentation I don't think that clang-docs will like the \brief and so on, I'll check this.

**Comment by daminetreg** [comment]
Note that the code is already in boost.asynchronous here : https://github.com/henry-ch/asynchronous/commit/5d9213743d22d6dcf5a33b84937fdb2be36faf1d#commitcomment-8883094

File: pre/traits/function_traits.hpp (hunk 3 above)
Comment (dan-42) [inline pre/traits/function_traits.hpp]: I liked to have `cpp  pre::function_traits`  but you are the architect of this library, so I'll change it of course.

Comment (daminetreg) [inline pre/traits/function_traits.hpp]: We can still make a convenience header with a using declaration, but I prefer things to be separated, as this library is kind of temporary library which should never grow and always go smaller as we should try to provide things to boost or remove them as soon as it's available in boost.

**Comment by dan-42** [comment]
@daminetreg I changed the namespace to `cpp  pre::lambda::function_traits`
and tests for captured lambdas has been added.
please note that a captured lambda has a different type as a std::function<>  so the comparing in the test checks for difference

File: test/CMakeLists.txt (hunk 1 above)
Comment (dan-42) [inline test/CMakeLists.txt:37]: well your refering hunter package does not support boost 1.59

**Comment by dan-42** [comment]
@daminetreg btw, does your hunter fork support lib-pre?
and Have you tried including boost 1.60 into hunter? did you need todo any changes?

**Comment by daminetreg** [comment]
Cool, Thanks for the changes :smile: I'll test and merge.

> @daminetreg btw, does your hunter fork support lib-pre?
> Yes
> 
> and Have you tried including boost 1.60 into hunter? did you need todo any changes?

No I have 1.59.0, but the change might be just increasing the version number if you don't need the new libraries added in 1.60. I mean adding a version in the file cmake/projects/Boost/hunter.cmake.

We should try to find budget to help hunter more, because it's a marvelous tool.

**Comment by daminetreg** [comment]
Ok we will be able to merge soon the wine is on it's way to Paris: 
img_20160206_112105

**Comment by dan-42** [comment]
awesome, let me know I you need/want any changes befoer you merge.

**Comment by dan-42** [comment]
@daminetreg any update on this one? 
cheers

File: test/CMakeLists.txt (hunk 1 above)
Comment (daminetreg) [inline test/CMakeLists.txt:37]: My locally cloned one does.

**Comment by daminetreg** [comment]
@dan-42 I updated a bit the code to support as well freestanding function and ensured the documentation get generated properly.

Can you take a look and tell me if it fits your needs as well ? I've also moved it to pre/type_traits, because we could brush it up to get it into boost as evolution of : 
http://www.boost.org/doc/libs/1_59_0/libs/type_traits/doc/html/boost_typetraits/reference/function_traits.html

**Comment by daminetreg** [comment]
I've also added support for member functions under the same API. Please tell me if you agree.

**Comment by dan-42** [comment]
Awesome I'll look into it this evening

**Comment by dan-42** [comment]
@daminetreg Hey this is cool! Thx for the changes.

I would be very happy to see this now merged :smile:

**Comment by daminetreg** [comment]
Great. :+1: :smile: Thanks for contributing this.

**Pull request merged**